crates/router/src/core/payment_methods/transformers.rs

```diff
@@ -1,104 +1,229 @@
+pub struct RetrieveCardResp {
     pub status: String,
+    pub error_message: Option<String>,
+    pub error_code: Option<String>,
+    pub payload: Option<RetrieveCardRespPayload>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct RetrieveCardRespPayload {
+    pub card: Option<Card>,
+    pub enc_card_data: Option<String>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct DeleteCardResp {
+    pub status: String,
+    pub error_message: Option<String>,
+    pub error_code: Option<String>,
+}
+
+pub fn basilisk_hs_key_id() -> &'static str {
+    "1"
 }
 
-pub fn mk_add_card_request(
+pub fn get_dotted_jwe(jwe: encryption::JweBody) -> String {
+    let header = jwe.header;
+    let encryption_key = jwe.encrypted_key;
+    let iv = jwe.iv;
+    let encryption_payload = jwe.encrypted_payload;
+    let tag = jwe.tag;
+    format!("{header}.{encryption_key}.{iv}.{encryption_payload}.{tag}")
+}
+
+pub fn get_dotted_jws(jws: encryption::JwsBody) -> String {
+    let header = jws.header;
+    let payload = jws.payload;
+    let signature = jws.signature;
+    format!("{header}.{payload}.{signature}")
+}
+
+pub async fn get_decrypted_response_payload(
+    jwekey: &Jwekey,
+    jwe_body: encryption::JweBody,
+) -> CustomResult<String, errors::VaultError> {
+    let public_key = encryption::KeyHandler::get_kms_decrypted_key(
+        jwekey,
+        jwekey.vault_encryption_key.to_string(),
+    )
+    .await
+    .change_context(errors::VaultError::SaveCardFailed)
+    .attach_printable("Fails to get public key of vault")?;
+    let private_key =
+        encryption::KeyHandler::get_kms_decrypted_key(jwekey, jwekey.vault_private_key.to_string())
+            .await
+            .change_context(errors::VaultError::SaveCardFailed)
+            .attach_printable("Error getting private key for signing jws")?;
+
+    let jwt = get_dotted_jwe(jwe_body);
+    let key_id = basilisk_hs_key_id();
+    let alg = jwe::RSA_OAEP;
+    let jwe_decrypted = encryption::decrypt_jwe(jwekey, &jwt, key_id, key_id, private_key, alg)
+        .await
+        .change_context(errors::VaultError::SaveCardFailed)
+        .attach_printable("Jwe Decryption failed for JweBody for vault")?;
+
+    let jws = jwe_decrypted
+        .parse_struct("JwsBody")
+        .change_context(errors::VaultError::ResponseDeserializationFailed)?;
+    let jws_body = get_dotted_jws(jws);
+
+    encryption::verify_sign(jws_body, &public_key)
+        .change_context(errors::VaultError::SaveCardFailed)
+        .attach_printable("Jws Decryption failed for JwsBody for vault")
+}
+
+pub async fn mk_basilisk_req(
+    jwekey: &Jwekey,
+    jws: &str,
+) -> CustomResult<encryption::JweBody, errors::VaultError> {
+    let jws_payload: Vec<&str> = jws.split('.').collect();
+    when(jws_payload.len().ne(&3), || {
+        Err(errors::VaultError::SaveCardFailed)
+    })?;
+    let jws_body = encryption::JwsBody {
+        header: jws_payload[0].to_owned(),
```

Comment:
We should not use direct indexing this cause panic's. Iterate and take values, if not found throw error.

Decoupling using `when(jws_payload.len().ne(&3)` solve that but the code still has a way to panic with this check remove or bypassed in future.

Reply:
sure